### PR TITLE
core: multinodes support for ket format, simple objects support, #533, #530, #523, #506, #413

### PR DESCRIPTION
**Pull request by even1024**
multinodes support for ket format, simple objects support, #533, #530, #523, #506, #413.

### CODE REVIEW

File: core/render2d/src/render_internal.cpp [hunk 1]
```diff
@@ -2005,6 +2008,28 @@ void MoleculeRenderInternal::_renderBonds()
...
+                break;
+                case KETSimpleObject::EKETRectangle:
+
+                break;
+                case KETSimpleObject::EKETLine:
+                break;
+            }
+        }
+    }*/
+}
```
Comment (mkviatkovskii) [inline core/render2d/src/render_internal.cpp:2031]: This looks like a dead code.

File: core/indigo-core/graph/graph.h [hunk 2]
```diff
@@ -253,6 +268,8 @@ namespace indigo
         bool _components_valid;
         int _components_count;
 
+        PtrArray<GraphMetaObject> _meta_data;
```
Comment (mkviatkovskii) [inline core/indigo-core/graph/graph.h:271]: Why do we use PtrArray? It should be considered as deprecated data structure now.

File: core/indigo-core/graph/graph.h [hunk 3]
```diff
@@ -24,6 +24,9 @@
 #include "base_cpp/non_copyable.h"
 #include "base_cpp/obj_array.h"
 #include "base_cpp/obj_pool.h"
+#include <list>
+#include <memory>
```
Comment (mkviatkovskii) [inline core/indigo-core/graph/graph.h:28]: Standard library headers should be included before local headers.

Comment (mkviatkovskii) [reply core/indigo-core/graph/graph.h:28]: BTW, I don't see any usage of list or memory here.

File: core/indigo-core/molecule/src/base_molecule.cpp [hunk 4]
```diff
@@ -278,6 +278,12 @@ void BaseMolecule::_mergeWithSubmolecule_Sub(BaseMolecule& mol, const Array<int>
         reaction_atom_exact_change[mapping[i]] = mol.reaction_atom_exact_change[i];
     }
 
+    if (skip_flags & COPY_BOND_DIRECTIONS)
+    {
+        _bond_directions.expandFill( edgeEnd() + mol.edgeEnd(), 0);
+    } else
+      _bond_directions.expandFill(  mol.edgeEnd(), 0);
+
```
Comment (mkviatkovskii) [inline core/indigo-core/molecule/src/base_molecule.cpp:286]: It's definitely required to run clang-format on changed files.

File: core/indigo-core/molecule/src/base_molecule.cpp [hunk 5]
```diff
@@ -4122,6 +4128,34 @@ const int* BaseMolecule::getPyramidStereocenters(int idx) const
     return stereocenters.getPyramid(idx);
 }
 
+void BaseMolecule::setStereoFlagPosition(int frag_index, const Vec3f& pos)
+{
+    try
+    {
+        _stereo_flag_positions.insert(frag_index, pos);
+    }
+    catch (...)
```
Comment (mkviatkovskii) [inline core/indigo-core/molecule/src/base_molecule.cpp:4137]: Why do we need such generic catch? It looks like a bad code.

File: core/render2d/src/render_internal.cpp (hunk 1 above)
Comment (even1024) [reply core/render2d/src/render_internal.cpp:2031]: fixed

File: core/indigo-core/graph/graph.h (hunk 2 above)
Comment (even1024) [reply core/indigo-core/graph/graph.h:271]: We can't use std containers here

File: core/indigo-core/graph/graph.h (hunk 3 above)
Comment (even1024) [reply core/indigo-core/graph/graph.h:28]: fixed

File: core/indigo-core/molecule/src/base_molecule.cpp (hunk 4 above)
Comment (even1024) [reply core/indigo-core/molecule/src/base_molecule.cpp:286]: fixed

File: core/indigo-core/molecule/src/base_molecule.cpp (hunk 5 above)
Comment (even1024) [reply core/indigo-core/molecule/src/base_molecule.cpp]: fixed

**Review by mkviatkovskii — APPROVED** [review]

**Pull request merged**